tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -2352,216 +2474,259 @@ private[optimizer] abstract class OptimizerCore(
+              PreTransLit(StringLiteral(r))) =>
+            PreTransBinaryOp(String_+, ll, PreTransLit(StringLiteral(lr + r)))
+          case (PreTransBinaryOp(String_+, PreTransLit(StringLiteral("")), lr), _) =>
+            PreTransBinaryOp(String_+, lr, rhs1)
           case _ =>
             stringDefault
       }
 
       case Int_+ =>
         (lhs, rhs) match {
-          case (IntLiteral(l), IntLiteral(r)) => IntLiteral(l + r)
-          case (_, IntLiteral(_))             => foldBinaryOp(Int_+, rhs, lhs)
-          case (IntLiteral(0), _)             => rhs
+          case (_, PreTransLit(IntLiteral(_))) =>
+            foldBinaryOp(Int_+, rhs, lhs)
+          case (PreTransLit(IntLiteral(0)), _) =>
+            rhs
 
-          case (IntLiteral(x),
-              BinaryOp(innerOp @ (Int_+ | Int_-), IntLiteral(y), z)) =>
-            foldBinaryOp(innerOp, IntLiteral(x+y), z)
+          case (PreTransLit(IntLiteral(x)),
+              PreTransBinaryOp(innerOp @ (Int_+ | Int_-),
+                  PreTransLit(IntLiteral(y)), z)) =>
+            foldBinaryOp(innerOp, PreTransLit(IntLiteral(x + y)), z)
 
-          case _                              => default
+          case _ => default
         }
 
       case Int_- =>
         (lhs, rhs) match {
-          case (_, IntLiteral(r)) => foldBinaryOp(Int_+, lhs, IntLiteral(-r))
+          case (_, PreTransLit(IntLiteral(r))) =>
+            foldBinaryOp(Int_+, lhs, PreTransLit(IntLiteral(-r)))
+
+          case (PreTransLit(IntLiteral(x)),
+              PreTransBinaryOp(Int_+, PreTransLit(IntLiteral(y)), z)) =>
+            foldBinaryOp(Int_-,
+                PreTransLit(IntLiteral(x - y)), z)
 
-          case (IntLiteral(x), BinaryOp(Int_+, IntLiteral(y), z)) =>
-            foldBinaryOp(Int_-, IntLiteral(x-y), z)
-          case (IntLiteral(x), BinaryOp(Int_-, IntLiteral(y), z)) =>
-            foldBinaryOp(Int_+, IntLiteral(x-y), z)
+          case (PreTransLit(IntLiteral(x)),
+              PreTransBinaryOp(Int_-, PreTransLit(IntLiteral(y)), z)) =>
+            foldBinaryOp(Int_+, PreTransLit(IntLiteral(x - y)), z)
 
-          case (_, BinaryOp(Int_-, IntLiteral(0), x)) =>
+          case (_, PreTransBinaryOp(Int_-, PreTransLit(IntLiteral(0)), x)) =>
             foldBinaryOp(Int_+, lhs, x)
 
           case _ => default
         }
 
       case Int_* =>
         (lhs, rhs) match {
-          case (IntLiteral(l), IntLiteral(r)) => IntLiteral(l * r)
-          case (_, IntLiteral(_))             => foldBinaryOp(Int_*, rhs, lhs)
+          case (_, PreTransLit(IntLiteral(_))) =>
+            foldBinaryOp(Int_*, rhs, lhs)
 
-          case (IntLiteral(1), _)  => rhs
-          case (IntLiteral(-1), _) => foldBinaryOp(Int_-, IntLiteral(0), rhs)
+          case (PreTransLit(IntLiteral(1)), _)  =>
+            rhs
+          case (PreTransLit(IntLiteral(-1)), _) =>
+            foldBinaryOp(Int_-, PreTransLit(IntLiteral(0)), rhs)
```

Comment:
Not in this PR, at least. That's a very unrelated change. I'll probably do that at the same time #2456 during my Scala Days talk.